doc/iterations/2024.3/shapes/configurable_collapse.adoc

```diff
@@ -0,0 +1,47 @@
+= Configurable collapse
+
+== Problem
+
+Sirius Web doesn't provide a way for specifiers to set the default collapsible state of a node. It also doesn't provide a way to let specifiers define tools that collapse/expand nodes as part of other operations (e.g. to expand a collapsed node when a child node is created).
+
+Nodes are always created expanded, and the only way to collapse them right now is that the end user does it manually.
+
+== Key Result
+
+- A specifier must be able to define NodeDescriptions that are collapsed/expanded by default.
+- A specifier can define tools that collapse/expand other nodes.
+
+== Solution
+
+A new field _Default Collapsing State Expression_ should be displayed in NodeDescription. The field contains an expression which is evaluated on each diagram render to collapse/expand a node.
+
+This expression would be used to set the Collapsing State if it hasn't been manually set by the end user or updated by a tool.
+
+A new operation _Set Collapsing State_ should be available in tools to expand/collapse nodes. This operation would have a _Graphical Candidates Expression_ field used to retrieve the nodes to expand/collapse, as well as a _CollapsingState_ field containing the value to set (Collapsed or Expanded).
+
+== Scenario
+
+1. Default Collapsing State
+- A specifier adds a NodeDescription in the view DSL
+- The _Default Collapsing State Expression_ field is available and its default value is empty, which means expanded to keep the existing default behavior of Sirius Web
+- The specifier can define an AQL expression to set the default Collapsing State of the node
+- When an end user creates a node represented in this manner in a diagram, the node is collapsed/expanded according to this expression.
+- If the user manually expands/collapses the node the default _Collapsing State Expression_ is not taken into account on the next renders.
+
+2. Collapse/Expand operation
+- A specifier creates a tool in the view DSL
+- The operation _Set Collapsing State_ is available in the "new" dropdown menu and creates a _Set Collapsing State_ element under the tool
+- The operation has two fields _Graphical Candidate Expression_ (a text field accepting an AQL expression) and _CollapsingState_ (a dropdown with 2 values: Collapsed and Expanded)
```

Comment:
Couldn't you do something like `aql: if('diagram.nodes->size() % 2 == 1) then diagram.nodes else Sequence{} endif`? I agree it is not pretty though, and in any case it would be easier/cleaner with a service.

Reply:
Yes you could indeed but if my use case becomes a tiny bit more complex with if it's an even number of node collapse everything or do something even more complicated you end up having to write complex stuff in AQL or even in model operations (you would end up having maybe two model operations and conditionally decide which one to call).

None of our advanced users wants to code in AQL or model operations.   (✖╭╮✖)